include/compiler.hpp

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+
+#include <vector>
+#include <memory>
+#include <mutex>
+#include <variant>
+#include <cstddef>
+
+class Logger
+{
+protected:
+    std::recursive_mutex &m_lock;
+    std::string &m_buildLog;
+
+public:
+    Logger(std::recursive_mutex &lock, std::string& build_log)
+        : m_lock(lock), m_buildLog(build_log)
+    {
+    }
+    void Log(const char *msg) const;
+};
+
+// An abstraction over a program binary
+class ProgramBinary
+{
+public:
+    struct Kernel
+    {
+        struct Arg
+        {
+            enum class AddressSpace
+            {
+                Private, Constant, Local, Global
+            };
+            const char *name;
+            const char *type_name;
+            bool readable, writable;
+            bool is_const, is_restrict, is_volatile;
+            AddressSpace address_qualifier;
+        };
+        enum class VecHintType
+        {
+            Char, Short, Int, Long, Half, Float, Double
+        };
+
+        const char *name;
+        std::vector<Arg> args;
+        unsigned vec_hint_size;
+        VecHintType vec_hint_type;
+    };
+
+    virtual ~ProgramBinary() = default;
+    virtual size_t GetBinarySize() const = 0;
+    virtual const void* GetBinary() const = 0;
+
+    const std::vector<Kernel> &GetKernelInfo() const;
+
+protected:
+    std::vector<Kernel> m_KernelInfo;
+};
+
+// An abstraction over DXIL + metadata
+class CompiledDxil
+{
+public:
+    struct Metadata
+    {
+        struct Arg
+        {
+            unsigned offset, size;
+
+            struct Image
+            {
+                unsigned buffer_ids[3];
+                unsigned num_buffer_ids;
+            };
+            struct Sampler
+            {
+                unsigned sampler_id;
+            };
+            struct Memory
+            {
+                unsigned buffer_id;
+            };
+            struct Local
+            {
+                unsigned sharedmem_offset;
+            };
+            std::variant<std::monostate, Image, Sampler, Memory, Local> properties;

```

Comment:
First time seeing `variant`, pretty cool :)

Reply:
It's not quite as nice as Rust's enums, but it's better than a raw enum + union combo.